src/DotVVM.Framework/Compilation/Binding/ExpressionHelper.cs

```diff
@@ -252,6 +259,53 @@ private static MethodRecognitionResult TryCallMethod(MethodInfo method, Type[] t
             };
         }
 
+        private static Type ResolveGenericTypeFromGivenExpressions(int i, Type genericArgument, ParameterInfo[] parameters, Expression[] args)
+        {
+            for (var j = 0; j < parameters.Length; j++)
+            {
+                var parameter = parameters[j];
+                if (parameter.ParameterType.IsGenericParameter && parameter.ParameterType.GenericParameterPosition == i)
+                {
+                    return args[i].Type;
+                }
+                if (parameter.ParameterType.IsArray)
+                {
+                    var elementType = parameter.ParameterType.GetElementType();
+                    var genericArgName = elementType.Name;
+                    var genericParameterPosition = elementType.GenericParameterPosition;
```

Comment:
I doubt this branch does anything. Could you please add a test that checks that arrays also work? For example call this method:

```csharp
T[] Flatten<T>(List<T[]> x)
```

Reply:
I have added support for this case. Also, I have added tests for it.